add support for CLI parameters for config and settings file

yauh opened the pull request:
Second try - now for mupx.
Added minimist to determine CLI parameters and allow users to override default `settings.json` and `mup.json`.

arunoda commented on lib/actions.js:
```diff
@@ -218,8 +222,8 @@ Actions.prototype.logs = function() {
 };
 
 Actions.init = function() {
```
I think it's better if we can directly get these filenames via the arguments.

yauh replied:
I'm not sure I understand. Could you clarify?

arunoda replied:
I mean passing name of tool files like this:

~~~js
Actions.init  = function(mupJsonFileName, settingFileName) {

}
~~~
Then we don't need to use `minimist` everywhere.

yauh replied:
you are absolutely right

arunoda commented on lib/config.js:
```diff
@@ -3,11 +3,13 @@ var path = require('path');
 
 require('colors');
 
 exports.read = function() {
```
May be we can provide the filename as the first argument

yauh replied:
Do you mean like `mup --config=mup2.json deploy`?
Since the filename is an optional argument I would put it second, but I am happy as long as it is there :)

arunoda replied:
Nope. I didn't mean that.
I mean this:

~~~js
exports.read = function(configFileName) {
..
}
~~~

yauh replied:
I see - fixed.

yauh commented:
Updated based on your comments.

yauh commented:
I wonder if eventually it may be best to add the value for `settingsfileName` directly inside the `mup.json` configuration instead of on the command line. Then all configuration is done from a central point.

rhyslbw commented:
@yauh I agree

arunoda commented:
Great. Mentioning the settings in the mup.json is a good idea. We'll think about that in a different PR. Let's work on this. 
In order to make this PR rock solid, we need to apply some of things like path rewrites as so on. So, I did it. Here's the patch. Try to include this.

~~~patch
diff --git a/bin/mup b/bin/mup
index b64e65f..4f2a91c 100755
--- a/bin/mup
+++ b/bin/mup
@@ -22,7 +22,7 @@ console.log('Configuration file : ' + configFileName);
 console.log('Settings file      : ' + settingsFileName);
 console.log('');
 
-var action = process.argv[2];
+var action = cliOptions._[0];
 if(action == 'init') {
   //special setup for init
   ActionsRegistry.init();
@@ -30,15 +30,15 @@ if(action == 'init') {
   var cwd = path.resolve('.');
   //read config and validate it
   var config = Config.read(configFileName);
-
-  // add settings file to configuration
-  config.settings = settingsFileName;
+  var runOptions = {
+    settingsFileName: settingsFileName
+  };
 
   if(config._passwordExists) {
     //check for sshpass
     helpers.checkSshPassExists(function(exists) {
       if(exists) {
-        runActions(config, cwd);
+        runActions(config, cwd, runOptions);
       } else {
         var errorMessage = "sshpass required for password based authentication: refer http://git.io/_vHbvQ";
         console.error(errorMessage.bold.red);
@@ -46,13 +46,13 @@ if(action == 'init') {
       }
     });
   } else {
-    runActions(config, cwd);
+    runActions(config, cwd, runOptions);
   }
 }
 
 
-function runActions(config, cwd, cliOptions) {
-  var actionsRegistry = new ActionsRegistry(config, cwd, cliOptions);
+function runActions(config, cwd, options) {
+  var actionsRegistry = new ActionsRegistry(config, cwd, options);
   if(actionsRegistry[action]) {
     actionsRegistry[action]();
   } else {
diff --git a/lib/actions.js b/lib/actions.js
index c9a64e2..01751c4 100644
--- a/lib/actions.js
+++ b/lib/actions.js
@@ -15,11 +15,11 @@ require('colors');
 
 module.exports = Actions;
 
-function Actions(config, cwd) {
+function Actions(config, cwd, options) {
   this.cwd = cwd;
   this.config = config;
   this.sessionsMap = this._createSessionsMap(config);
-  this.settingsFileName = config.settings;
+  this.settingsFileName = options.settingsFileName;
 
   //get settingsFileName into env
   var setttingsJsonPath = path.resolve(this.cwd, this.settingsFileName);
diff --git a/lib/config.js b/lib/config.js
index bf936ff..1da573b 100644
--- a/lib/config.js
+++ b/lib/config.js
@@ -8,6 +8,10 @@ require('colors');
 
 exports.read = function(configFileName) {
   var mupJsonPath = path.resolve(configFileName);
+  // path of the mup.json file is the basedir and everything 
+  // will be based on that path
+  var basedir = path.dirname(mupJsonPath);
+
   if(fs.existsSync(mupJsonPath)) {
     var mupJson = cjson.load(mupJsonPath);
 
@@ -21,7 +25,7 @@ exports.read = function(configFileName) {
     if(typeof mupJson.setupPhantom === "undefined") {
       mupJson.setupPhantom = true;
     }
-    mupJson.meteorBinary = (mupJson.meteorBinary) ? getCanonicalPath(mupJson.meteorBinary) : 'meteor';
+    
     if(typeof mupJson.appName === "undefined") {
       mupJson.appName = "meteor";
     }
@@ -49,7 +53,7 @@ exports.read = function(configFileName) {
 
         if(server.pem) {
           server.pem =
-            rewriteHome(server.pem, "SSH private key file is invalid");
+            rewritePath(server.pem, "SSH private key file is invalid");
         } else {
           //hint mup bin script to check whether sshpass installed or not
           mupJson._passwordExists = true;
@@ -65,51 +69,53 @@ exports.read = function(configFileName) {
     }
 
     //rewrite ~ with $HOME
-    mupJson.app = rewriteHome(
+    mupJson.app = rewritePath(
       mupJson.app, "There is no meteor app in the current app path.");
 
     if(mupJson.ssl) {
       mupJson.ssl.port = mupJson.ssl.port || 443;
-      mupJson.ssl.certificate = rewriteHome(
+      mupJson.ssl.certificate = rewritePath(
         mupJson.ssl.certificate, "SSL certificate file does not exists.");
-      mupJson.ssl.key = rewriteHome(
+      mupJson.ssl.key = rewritePath(
         mupJson.ssl.key, "SSL key file does not exists.");
     }
 
     return mupJson;
   } else {
-    console.error('configuration file ' + configFileName + ' does not exist!'.red.bold);
+    var messgae = 
+      'configuration file ' + configFileName + ' does not exist!'.red.bold
+    console.error(messgae);
     helpers.printHelp();
     process.exit(1);
   }
-};
+  
+  function rewritePath(location, errorMessage) {
+    if(!location) {
+      return mupErrorLog(errorMessage);
+    }
 
-function rewriteHome(location, errorMessage) {
-  if(!location) {
-    return mupErrorLog(errorMessage);
-  }
+    var location = location.replace('~', process.env.HOME).trim();
+    if(location.indexOf(0) !== "/" || location.indexOf(0) !== "\\") {
+      // if path start with / or \ (on windows)
+      // we need to make sure, they are from the basedir
+      // but not from the current dir
+      location = path.resolve(basedir, location);
+    } else {
+      // anyway, we need to resolve path for windows compatibility
+      location = path.resolve(location);
+    }
 
-  var location = location.replace('~', process.env.HOME);
-  location = path.resolve(location);
+    if(!fs.existsSync(location)) {
+      mupErrorLog(errorMessage);
+    }
 
-  if(!fs.existsSync(location)) {
-    mupErrorLog(errorMessage);
+    return location;
   }
 
-  return location;
-}
-
-function mupErrorLog(message) {
-  var errorMessage = 'Invalid configuration file ' + configFileName + ': ' + message;
-  console.error(errorMessage.red.bold);
-  process.exit(1);
-}
-
-function getCanonicalPath(location) {
-  var localDir = path.resolve(__dirname, location);
-  if(fs.existsSync(localDir)) {
-    return localDir;
-  } else {
-    return path.resolve(rewriteHome(location));
+  function mupErrorLog(message) {
+    var errorMessage = 'Invalid configuration file ' + configFileName + ': ' + message;
+    console.error(errorMessage.red.bold);
+    process.exit(1);
   }
-}
+};
+

~~~

yauh commented:
added your changes. Fixed the spelling for `messgae` and var reference for `basedir`.
I agree that moving settings to mup.json is a completely different PR.

yauh commented:
I just saw this PR duplicates https://github.com/arunoda/meteor-up/pull/337/

arunoda commented:
Okay. 
Released with `mupx@1.3.0`

yauh commented:
Super, thanks! That will be quite helpful for some of my deployments :+1:
